### PR TITLE
Blueprint issues

### PR DESCRIPTION
**Pull request by pvliss**
This PR contains the following fixes:
1. Fix to not check for blueprint existence when `skipChecks` is true to allow for running mocha tests in blueprints as well, even if the npm module is no linked 
2. Fix to pass `options` and `configOptions` to blueprint generators and sub-generators
3. Add mocha tests for `spring-controller` and `service-blueprints` generators

I have tested(meaning `npm test`) these changes in the generator and also with the Kotlin and VueJS blueprints

Relates to issue 1 of #9232 

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by pvliss** [comment]
Closing and re-opening to trigger a CI rebuild as I do not think that the e2e failure is related to this PR

**Comment by murdos** [comment]
I agree with 2. and 3. and think this is a nice improvement, but I don't understand why 1. is needed.
Could you explain it? I don't remember having issues with mocha tests when not linking the npm module.

**Comment by pvliss** [comment]
Hello @murdos and thanks for reviewing this.

So when making these changes other than testing them against the Kotlin blueprint I also wanted to test against another blueprint just to be safe. So I decided to use jhipster-vuejs which I know is currently really active :).  So I forked it and then tried to run the mocha tests using `npm test` without linking or having previously installed the module. The problem is that method `checkBlueprint` would try to find the jhipster-vuejs within `node_modules` but since it was not there it was failing even though the module was added as a dependency in the mocha tests. I could have solved this problem by just linking the jhipster-vuejs but at the time I had run too many tests too many times and that was my workaround.

So to sum it up it is really not that important and can be resolved by running `npm link` hence I can revert these changes if you wish.

**Comment by murdos** [comment]
Alright. I'll take a look to vuejs blueprint to see how the tests are configured, and I'll let you know.

**Comment by pvliss** [comment]
Thanks @murdos . Just to note that this is not specific to the VueJS blueprint. The Kotlin blueprint and most probably any blueprint has the same issue

**Comment by pascalgrimaud** [comment]
thanks a lot @pvliss and @murdos
We plan to do a new release of Vue.js blueprint next week, probably monday with @sahbi-ktifa : let us know if you see any issue with blueprint / tests

**Comment by murdos** [comment]
I checked what I did on my own blueprint for work, and the use of `skipChecks` makes sense and is a good idea! Good job :)
I was not convinced at beginning, because when you write unit test for a single subgenerator (that doesn't call other subgen), it's not useful (that's mainly what I did on my own blueprint). But as soon as you run unit tests on the `app` generator, this is very useful.
In fact I used a different technique to do that, but it's very loud and can't be redone on each blueprint: https://github.com/jhipster/generator-jhipster/blob/master/test/blueprint/app-blueprint.spec.js#L51 => could you check if this can be removed with your change (I guess so) and same thing for the entity-blueprint.spec.js ? Thanks

Last question: should we fix that on V5 or on V6 of the generator? If V6 is released in weeks or months, I'd opt for V5.

**Comment by pvliss** [comment]
> I checked what I did on my own blueprint for work, and the use of `skipChecks` makes sense and is a good idea! Good job :)
> I was not convinced at beginning, because when you write unit test for a single subgenerator (that doesn't call other subgen), it's not useful (that's mainly what I did on my own blueprint). But as soon as you run unit tests on the `app` generator, this is very useful.

Thanks. Basically my issue was that the blueprint had to be installed or linked in order to run the unit tests which should not be the case as the `withGenerators` call of `yeoman-test` is what this is all about. Not sure if the optimal solution though.

> In fact I used a different technique to do that, but it's very loud and can't be redone on each blueprint: https://github.com/jhipster/generator-jhipster/blob/master/test/blueprint/app-blueprint.spec.js#L51 => could you check if this can be removed with your change (I guess so) and same thing for the entity-blueprint.spec.js ? Thanks

You mean remove all lines from 51-56? If we do that then the test will fail I think? From what I understand it is just mocking the presence of a blueprint even if it does nothing and simply asserts the existence of a blueprint. Am I missing something? Will test anyways

**Comment by pvliss** [comment]
@murdos You were right. It works!!! But I think that this does not make the test on its own very useful, does it? Maybe I should add few more tests that actually assert something more useful, e.g. creating some files while skipping others, etc. WDYT?

**Comment by murdos** [comment]
This test has one purpose: make sure the blueprint version is saved in .yo-rc.json and that the npm dependency is added to package.json, so in my opinion it is useful as it is.
Then each jhipster generator has its own blueprint test, to make sure it works as a blueprint.

Could you remove all lines from 51-56, do the same in `entity-blueprint.spec.js`, and also remove the `createBlueprintMockForSubgen` function in utils.js? Thanks :)

**Comment by pvliss** [comment]
@murdos Done. Run locally `npm test` and everything passed. Let wait for the CI as well. Thanks again for the review.
